### PR TITLE
Support running on demand via systemd

### LINKED ISSUE #126: Support starting rest-server from inetd and similar services
Output of `rest-server --version`
---------------------------------
rest-server 0.10.0 compiled with go1.14.10 on illumos/amd64

What should rest-server do differently?
---------------------------------------

At the moment the rest-server must be either run as root to bind to low ports like 443 or use a high port to be able to run unprivileged. Being able to use 443 is desirable, having to run as root is not.

What are you trying to do? What is your use case?
-------------------------------------------------

Integration with inetd(1) solves the problem with minimal complexity in the application. Other service managers generally provide compatibility, e.g. SMF on Solaris/Illumos.

### PR DESCRIPTION
**Pull request by fd0**
What is the purpose of this change? What does it change?
--------------------------------------------------------

This change allows `rest-server` to be run on demand by systemd (socket activation). The rationale for supporting both is as follows:

 1. On demand running with a listening socket opened externally is a good idea because then `rest-server` does not have to have the privileges to listen on a port.
 2. My first choice for that would be systemd (so the process can even be restricted further), but it is not available for non-Linux systems

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Closes #126 
Supersedes #127

Checklist
---------

- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/rest-server/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/rest-server/commits/master)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: cmd/rest-server/main.go [hunk 1]
```diff
@@ -111,21 +115,59 @@ func runRoot(cmd *cobra.Command, args []string) error {
...
+		return net.FileListener(os.Stdin)
+	}
+
+	// try systemd socket activation
+	listeners, err := activation.Listeners()
+	if err != nil {
+		panic(err)
+	}
+
+	switch len(listeners) {
```
Comment (wojas) [inline cmd/rest-server/main.go:146]: Perhaps make the systemd socket activation explicit with a `--systemd` switch?

Comment (fd0) [reply cmd/rest-server/main.go]: Why do you think that's needed? We can discover (from the environment) if we're run via systemd socket activation...

Comment (wojas) [reply cmd/rest-server/main.go]: In that case you will end up ignoring the address configuration and using systemd instead, which could surprise the user. Adding a `--systemd` flag also makes clear that some magic is happening here.

File: cmd/rest-server/listener_windows.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	"fmt"
+	"log"
+	"net"
+	"os"
+)
+
+// findListener creates a listener.
+func findListener(inetd bool, addr string) (listener net.Listener, err error) {
+	if inetd {
+		log.Print("inetd mode activated")
```
Comment (wojas) [inline cmd/rest-server/listener_windows.go:13]: Can this even work on windows?

File: cmd/rest-server/main.go (hunk 1 above)
Comment (wojas) [reply cmd/rest-server/main.go]: On the other hand, rest-server does log that it is in systemd mode, so I guess this is fine,

**Review by wojas — APPROVED** [review]
LGTM!

File: cmd/rest-server/listener_windows.go (hunk 2 above)
Comment (fd0) [reply cmd/rest-server/listener_windows.go:13]: Via cygwin maybe? Or WSL? It compiled, so I left it there...

**Comment by fd0** [comment]
inetd mode is off: https://github.com/restic/rest-server/issues/126#issuecomment-900565603

File: cmd/rest-server/listener_unix.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		return listener, nil
+
+	case 1:
+		// one listener supplied by systemd, use that one
+		//
+		// for testing, run rest-server with systemd-socket-activate as follows:
+		//
+		//    systemd-socket-activate -l 8080 ./rest-server
+		log.Printf("systemd socket activation mode")
+		return listeners[0], nil
```
Comment (jsonn) [inline cmd/rest-server/listener_unix.go:40]: I'm not sure who should do the check, but if the environmental flags don't match the passed in file descriptors, `listerners[0]` can be nil here and result in a panic later. Not a very nice to debug.

Comment (MichaelEischer) [reply cmd/rest-server/listener_unix.go:40]: Checking that `listeners[0]` is valid is probably a good idea before returning it. The expected behavior of `findListener` is that it either returns a listening socket or an error. Returning `nil` is thus wrong. Would you be willing to open a PR to check that `listeners[0] != nil`?

Comment (jsonn) [reply cmd/rest-server/listener_unix.go:40]: I've opened https://github.com/coreos/go-systemd/issues/405 for now, let's see if they are willing to fix the condition on their end.

**Pull request merged**